File: cpp/test_hash_table.cpp
#include <iostream>
#include <fstream>
#include <filesystem>
#include <thread>
#include <vector>
#include <string>
#include <cassert>
#include <random>

#include "hash_table.h"

std::random_device rd;
std::mt19937 gen(rd());
std::uniform_int_distribution<int> dis(INT_MIN, INT_MAX);

void assertCondition(bool condition, const std::string& message) {
    if (!condition) {
        std::cerr << "Assertion failed: " << message << std::endl;
        std::exit(EXIT_FAILURE);
    }
}

#define ASSERT(condition) assertCondition((condition), #condition)

void runTest(void (*testFunc)(), const std::string& testName) {
    std::cout << "Running " << testName << "..." << std::endl;
    testFunc();
    std::cout << testName << " passed." << std::endl;
}

int randint32() {
    return dis(gen);
}

void cleanDirectory(const std::string& dir) {
    std::filesystem::remove_all(dir);
}

void testSetup() {
    std::string tableDir = "_test_setup";
    std::filesystem::remove_all(tableDir);

    PersistentHashTable ht(tableDir);
    ASSERT(std::filesystem::exists(tableDir));
    ASSERT(std::filesystem::is_directory(tableDir));
    auto logFile = std::filesystem::path(tableDir) / "log.txt";
    ASSERT(std::filesystem::exists(logFile));

    std::filesystem::remove_all(tableDir);
}

void testCorrectness() {
    std::string tableDir = "_test_correctness";
    std::filesystem::remove_all(tableDir);

    PersistentHashTable ht(tableDir);
    const int n = 100;
    for (int i = 0; i < n; ++i) {
        int key = randint32();
        int value = randint32();
        ht.setItem(key, value);
        int retrieved = ht.getItem(key);
        ASSERT(retrieved == value);
    }

    std::filesystem::remove_all(tableDir);
}

void testReload() {
    std::string origDir = "_test_reload_orig";
    std::string reloadDir = "_test_reload_new";
    cleanDirectory(origDir);
    cleanDirectory(reloadDir);

    // initial hash table
    PersistentHashTable origHt(origDir);
    int key = randint32(), value = randint32();
    origHt.setItem(key, value);

    // reload hash table from log
    std::filesystem::path log_file = origHt.getLogFile();
    PersistentHashTable reloadedHt = PersistentHashTable(log_file, reloadDir, 3);

    // verify the reloaded data
    ASSERT(reloadedHt.getItem(key) == value);

    cleanDirectory(origDir);
    cleanDirectory(reloadDir);
}

void threadTask(PersistentHashTable* ht, int key, int value) {
    ht->setItem(key, value);
    int retrieved = ht->getItem(key);
    ASSERT(retrieved == value);
}

void testThreadSafety() {
    std::string tableDir = "_test_thread_safety";
    cleanDirectory(tableDir);

    PersistentHashTable ht(tableDir);
    const int n = 100; // number of threads

    std::vector<std::thread> threads;
    for (int i = 0; i < n; ++i) {
        threads.emplace_back(threadTask, &ht, i, randint32());
    }

    for (auto& t : threads) {
        t.join();
    }

    cleanDirectory(tableDir);
}



int main() {
    runTest(testSetup, "Test Setup");
    runTest(testCorrectness, "Test Correctness");
    // runTest(testSetError, "Test Error Handling");
    runTest(testReload, "Test Reload");
    runTest(testThreadSafety, "Test Thread Safety");

    std::cout << "All tests passed." << std::endl;
    return 0;
}

File: python/test_hash_table.py
import shutil
import threading
from random import randint
from pathlib import Path

from hash_table import PersistentHashTable


def randint32():
    return randint(-1 << 31, (1 << 31) - 1)


def test_setup():
    """
    Test setup of hash-table including log file
    """
    table_dir = Path('_test_setup')
    shutil.rmtree(table_dir, ignore_errors=True)

    # creation of new hash table
    ht = PersistentHashTable(table_dir)
    assert table_dir.exists()
    assert table_dir.is_dir()
    assert ht.log_file == table_dir / 'log.txt'
    assert ht.log_file.exists()
    del ht

    # instantiate existing hash table
    ht = PersistentHashTable(table_dir)
    assert table_dir.exists()
    assert table_dir.is_dir()
    assert (table_dir / 'log.txt').exists()
    assert ht.log_file == table_dir / 'log.txt'
    assert ht.log_file.exists()

    shutil.rmtree(table_dir)


def test_correct(n=5000):
    """
    Test correctness of hash-table set and get methods
    """
    table_dir = Path('_test_runtime')
    shutil.rmtree(table_dir, ignore_errors=True)

    ht = PersistentHashTable(table_dir)
    keys = [randint32() for _ in range(n)]
    vals = [randint32() for _ in range(n)]

    # set values
    for k, v in zip(keys, vals):
        ht[k] = v

    # assert correct storage
    for k, v in zip(keys, vals):
        assert ht[k] == v

    # assert correct logging
    assert len(ht.log_file.read_text().strip().split()) == n

    shutil.rmtree(table_dir)


def test_set_error(n=10):
    """
    Test the handling of errors during set method
    """
    table_dir = Path('_test_error')
    shutil.rmtree(table_dir, ignore_errors=True)

    ht = PersistentHashTable(table_dir)
    keys = [randint32() for _ in range(n)]
    vals = [randint32() for _ in range(n)]

    # forced error in setting
    try:
        ht[set()] = 'error'
    except TypeError:
        pass

    for k, v in zip(keys, vals):
        ht[k] = v

    # forced error in setting
    try:
        ht[set()] = 'error'
    except TypeError:
        pass

    # assert correct storage
    for k, v in zip(keys, vals):
        assert ht[k] == v

    # assert correct logging
    assert len(ht.log_file.read_text().strip().split()) == n

    shutil.rmtree(table_dir)


def test_reload(n=1000):
    """
    Test the reload of the hash table from a log file
    """
    orig_dir = Path('_test_reload_a')
    new_dir = Path('_test_reload_b')
    shutil.rmtree(orig_dir, ignore_errors=True)
    shutil.rmtree(new_dir, ignore_errors=True)

    orig_ht = PersistentHashTable(orig_dir)
    keys = [randint32() for _ in range(n)]
    vals = [randint32() for _ in range(n)]
    for k, v in zip(keys, vals):
        orig_ht[k] = v

    # assert correct hash table key-values
    new_ht = PersistentHashTable.from_log(orig_ht.log_file, new_dir)
    for k, v in zip(keys, vals):
        assert new_ht[k] == v

    # assert correct logging
    assert len(new_ht.log_file.read_text().strip().split()) == n

    shutil.rmtree(orig_dir)
    shutil.rmtree(new_dir)


def test_thread(n=300):
    """
    Test the thread safety of the hash table
    """
    table_dir = Path('_test_thread')
    shutil.rmtree(table_dir, ignore_errors=True)

    ht = PersistentHashTable(table_dir, bucket_mag=1)
    keys = [randint32() for _ in range(n)]
    vals = [randint32() for _ in range(n)]

    def set_key_val(ht, key, val):
        ht[key] = val
        assert ht[key] == val

    threads = []
    for i in range(n):
        t = threading.Thread(target=set_key_val, args=(ht, i, i))
        threads.append(t)
        t.start()

    for t in threads:
        t.join()

    # assert final result correctness
    for i in range(i):
        assert ht[i] == i

    # assert correct logging
    assert len(ht.log_file.read_text().strip().split()) == n

    shutil.rmtree(table_dir, ignore_errors=True)


if __name__ == '__main__':
    test_setup()
    test_correct()
    test_set_error()
    test_reload()
    test_thread()
